### PR TITLE
Comments block: Change ID to `core/comments`

### PR DESCRIPTION
**Pull request by cbravobernal**
Left as a draft to have some feedback about the rename first, the options we have are:

- Leave it as is ( is not a Query Loop yet)
- Rename it (inconsistency between ID - class,  and the rest of the name on the interface, folders, etc). We cannot update the block name, as has been stable a couple of months.

## What?

Renaming `Comments Query Loop` with `Comments`

## Why?

Right now, Comments Query Loop is not working like a Query block, it has not any option and is just rendering the comments, being able to add styling to them.

## How?

Search and replace!

## Testing Instructions

- On `trunk`, edit the "Single" template in the Site Editor to add the "Comments" block.
- Change some of the block's properties, e.g. text color or font size.
- Switch to the Code Editor and verify that the block's ID is `wp:comments-query-loop`. Copy-paste the block attributes, and the outermost `<div />` tag (including its HTML attributes) to a text editor of your choice; we'll need them later.
- Switch back to the Visual Editor.
- Switch to this branch (`update/rename-comments-query-loop`).
- Reload the template in the Site Editor.
- Open your browser console, and verify that it contains an "Updated blocks: core/comments` message.
- Switch to the Code Editor. Verify that the block ID is now `wp:comments` there ✅  Compare the block attributes and the outermost `<div />` tag to the ones you copied earlier. They should be identical, except now there should be an additional `wp-block-comments-query-loop` class (for backwards compatibility).
- Switch back to the Visual Editor.
- Verify that the block still works as before.
- Remove the block, and insert a new "Comments" block. Verify in the browser console that it doesn't trigger a block deprecation.
- Switch to the Code Editor, and verify that it _doesn't_ have the `wp-block-comments-query-loop` class.

Finally, try the same in the Post Editor.

### CODE REVIEW

**Comment by ockham** [comment]
> Right now, Comments Query Loop is not working like a Query block, it has not any option and is just rendering the comments, being able to add styling to them.

I'd add that it also has inner blocks that go beyond a query, such as the Comments Title, and the Comments Form (which we're going to include per #40256) 😄

File: packages/block-library/src/comments/block.json
Comment (ockham) [inline packages/block-library/src/comments/block.json]: Let's actually also change the ID:

```suggestion
	"name": "core/comments",
```

Reasoning: It's overall much less confusing to have the ID reflect the block name (and file and folder names), especially if there are plenty of related blocks with similar names.

As for the reasoning _why_ I think we should rename the block overall, see https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1105176912.
Let's add some logic to convert from the old to the new ID here: https://github.com/WordPress/gutenberg/blob/47a535914907153c62086f0d7bea521c628eca53/packages/blocks/src/api/parser/convert-legacy-block.js#L58-L68

Comment (ockham) [reply packages/block-library/src/comments/block.json]: Adding that it seems like the stabilized Comments Query Loop block has so far only been in on GB plugin release (v13.0.0): https://github.com/WordPress/gutenberg/commit/6f7c5b573a6023c2f2ce3eecd0d6b169ce4e964c

Bildschirmfoto 2022-04-21 um 15 07 12

**Comment by ockham** [comment]
Given latest developments (https://github.com/WordPress/gutenberg/issues/40521), I'm afraid we can close this now 😅

**Comment by ockham** [comment]
> Given latest developments (#40521), I'm afraid we can close this now 😅

Since we've decided to punt #40521, we're revisiting this PR 😬

**Comment by ockham** [comment]
Inserting a Comments Query Loop on `trunk`, and then switching to this branch is currently causing a block validation error:

image

Console:

image


I think we have to update a few `parent` and `ancestor` fields in some child blocks' `block.json` files (but even that doesn't seem to be quite enough to get rid of the validation error).

**Comment by ockham** [comment]
If we get totally stuck on the block validation error, our fallback could probably be to leave the block ID unchanged (i.e. keep it as `core/comments-query-loop`). I'd rather avoid that TBH, so it'd be good if we could try to fix the error for a bit longer, but if all else fails, we have at least a fallback strategy I guess 😅

File: packages/block-library/src/comments/index.js
Comment (luisherranz) [inline packages/block-library/src/comments/index.js]: I added a deprecation with the old class name to bypass the validation error. I've read a few times the [deprecation guide](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/), but I still don't know if this is a use case that is supposed to be handled like this. In a sense, I added a `save` with the old class name to say "hey Gutenberg, this change is expected and the attributes didn't change, so just update the HTML and don't bother the user with a warning message".

File: packages/block-editor/src/hooks/generated-class-name.js
Comment (luisherranz) [inline packages/block-editor/src/hooks/generated-class-name.js]: But the old `save` function added to the `deprecated` doesn't work yet by itself because this function (called in the `blocks.getSaveContent.extraProps` filter) is adding the `wp-block-comments` class name so we need to find a solution for that. _This is just a quick fix to show that it validates properly if this is bypassed._

**Comment by ockham** [comment]
Thanks @luisherranz! A deprecation/migration could be one promising strategy.

I'm still wondering, why are we getting the block invalidation error, when the [legacy block conversion](https://github.com/WordPress/gutenberg/blob/b8aa52dc4a1fe3887be27d5f3d1a77b409dc8750/packages/blocks/src/api/parser/convert-legacy-block.js) seems to be working for other blocks? I.e. what are they doing differently? (This is what lead me to assume that the problem here is that we have child blocks -- but maybe I'm mistaken?)

File: packages/block-library/src/comments/index.js
Comment (ockham) [reply packages/block-library/src/comments/index.js]: We might not need the deprecation (but can add the relevant logic to the Comments block's `save.js` instead?)

File: packages/block-editor/src/hooks/generated-class-name.js
Comment (ockham) [reply packages/block-editor/src/hooks/generated-class-name.js]: > called in the `blocks.getSaveContent.extraProps` filter

Can we add our own filter to that hook in order to remove the old class name? 🤔

Comment (ockham) [reply packages/block-editor/src/hooks/generated-class-name.js]: Oooh, or maybe we can solve our issue using [this filter](https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#blocks-getblockdefaultclassname)?

Comment (ockham) [reply packages/block-editor/src/hooks/generated-class-name.js]: > > called in the `blocks.getSaveContent.extraProps` filter
> 
> Can we add our own filter to that hook in order to remove the old class name? 🤔

```js
function setCommentsBlockCustomClassName( props ) {
	let { className } = props;
	const classes = className?.split( ' ' );
	if ( classes.includes( 'wp-block-comments' ) ) {
		className = classes
			.filter( ( c ) => c !== 'wp-block-comments-query-loop' )
			.join( ' ' );
	}
	return { ...props, className };
}

addFilter(
	'blocks.getSaveContent.extraProps',
	'core/comments/set-block-custom-class-name',
	setCommentsBlockCustomClassName
);
```

image


Well, not quite there yet 😅

**Comment by michalczaplinski** [comment]
I've figured out the problem and the reason was that in our case Gutenberg expects the `apiVersion` to be defined in the deprecations (!).


https://user-images.githubusercontent.com/5417266/165649615-54f7f959-ca7a-4e3e-800d-d55240594080.mp4


I think that this is not documented anywhere and is also very counter-intuitive.

**Comment by luisherranz** [comment]
Well, the API version of the deprecated block is 2, so this makes sense. It's true that in this case, we would be using it to bypass those filters, but it still seems correct because it's explicit in the conditional that those filters shouldn't run for block types with API versions bigger than 1.  

I agree it'd be great to get confirmation from someone with more expertise, though 🙂

I guess another question is, if we bypass all those filters, what happens with the other functionality? For example, are now old blocks with [custom class names](https://github.com/WordPress/gutenberg/blob/f297b6129b2250d479e0311c91a1f8dfd2092767/packages/block-editor/src/hooks/custom-class-name.js#L154-L174) triggering a validation error? _Although it's possible that Gutenberg is dealing with that logic somewhere else because if not the same would happen to all the API > 1 blocks._

File: packages/block-library/src/comments/index.js
Comment (luisherranz) [reply packages/block-library/src/comments/index.js]: > add the relevant logic to the Comments block's save.js

What logic? (sorry I'm not following 😄)

**Comment by ockham** [comment]
To recap:

The goal of this PR is to change the name (and ID) of the "Comments Query Loop" (`core/comments-query-loop`) block to (just) "Comments (`core/comments`).

We've added [logic to `convert-legacy-blocks.js`](https://github.com/WordPress/gutenberg/pull/40506/files#diff-6bf92c92e19be36ebe21679457ca6252639581f31d12e8f0b49326081c1dc1a0) in order to change the ID. However, [we're  still getting a block validation error](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1109880738):

image

The reason seems to be that we `useBlockProps.save()` on the wrapping HTML element in our `save.js`:

https://github.com/WordPress/gutenberg/blob/f297b6129b2250d479e0311c91a1f8dfd2092767/packages/block-library/src/comments/save.js#L6-L12

We currently work around this by adding a deprecation (which also wasn't straight-forward; we had to [set `apiVersion` to `2` for it to make sure the `blocks.getSaveContent.extraProps` filter is applied](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1111586444)).

To me, it seem wrong that on top of the legacy block conversion logic, we _also_ need a deprecation; I also think that it'll be easy to miss for devs that will work with this code in the future.

I feel like that using `useBlockProps.save()` on a block's wrapping element should be wide-spread enough to be properly handled by `convert-legacy-blocks.js`. Or maybe there's already a filter or something that we could use that I'm currently missing?

I'd be curious to hear your thoughts @gziolo @youknowriad

**Comment by youknowriad** [comment]
Hey @ockham to be honest, I'm not sure why you have to add a deprecation while with all the other previous case of legacyConvert, we didn't AFAIK. That said, this "legacyConvert" has always been a hack and not an official API, in that sense, it's possible that it doesn't work properly in all cases.

I'm curious to try to understand why this block is different than the previous legacy conversion we did.

**Comment by ockham** [comment]
> Hey @ockham to be honest, I'm not sure why you have to add a deprecation while with all the other previous case of legacyConvert, we didn't AFAIK.

Heh, that's exactly what puzzled me as well 😅  I can't say I've got a totally conclusive answer, but IIUC, the main difference is the `useBlockProps.save()` on the wrapping HTML element. From a quick look, the other blocks that are handled by `convert-legacy-blocks.js` don't do that.

What I _think_ happens is that when the block is parsed for validation, the previous block class name (`wp-block-comments-query-loop`) on the `div` is interpreted as a custom CSS class name by `fixCustomClassname` and kept, while the new one (`wp-block-comments`) is added.

> That said, this "legacyConvert" has always been a hack and not an official API, in that sense, it's possible that it doesn't work properly in all cases.

Yeah, it did look like a possbile shortcoming to me...

**Comment by luisherranz** [comment]
> What I think happens is that when the block is parsed for validation, the previous block class name (wp-block-comments-query-loop) on the div is interpreted as a custom CSS class name by fixCustomClassname and kept, while the new one (wp-block-comments) is added.

That is indeed happening, but bypassing `fixCustomClassname` didn't solve the validation error because it was still comparing:

- original content: `<div class="wp-block-comments-query-loop">`
- new content (new `save.js` output): `<div class="wp-block-comments">`

I didn't debug the `convert-legacy-blocks.js` code, but if it was supposed to take care of the renaming of that class, it was not working.

**Comment by youknowriad** [comment]
Paragraph and Cover blocks also use `useBlockProps.save` but these are the oldest conversions so maybe something changed in the validation/deprecation... front in the meantime, the other blocks seem all "dynamic blocks" so there's no need for deprecations there anyway.

Maybe a (hacky) way to solve this is to actually replace the classname in "originalContent" or "innerHTML" or whatever is used to check deprecation inside `convertLegacyBlocks` function.

**Comment by luisherranz** [comment]
I made a video to explain when the additional classes that make the validation fail are added, because it's different for the case of the new `save` (`fixCustomClassname`) than the deprecated `save` (`addGeneratedClassName`).

<a href="https://www.loom.com/share/5271810771f14c3a8ee607afe6c86158">
    
  </a>

https://www.loom.com/share/5271810771f14c3a8ee607afe6c86158

[Excalidraw](https://excalidraw.com/#json=LvRUO_5_DCDPdDDYW0upx,0z3FRUAB5wgI4vyenLRKvQ)

**Comment by ockham** [comment]
Thanks @luisherranz and @youknowriad for your thoughts!

At this point I'm inclined to give up on finding the "perfect" solution and just go with the deprecation 😬 

> Maybe a (hacky) way to solve this is to actually replace the classname in "originalContent" or "innerHTML" or whatever is used to check deprecation inside `convertLegacyBlocks` function.

I've thought about it, but I'm a bit worried about the ripple effects: The ID rename logic is happening in `convertLegacyBlockNameAndAttributes`, which is then wrapped in `convertLegacyBlocks` for the parser; but it's also called directly from `synchronizeBlocksWithTemplate`. So it seems we'd have to change the `originalContent` or `innerHTML` in both call sites; and I'm a bit worried that that could have some unwanted consequences (especially since I still don't have a very clear grasp of the order that functions are called during validation and deprecation).

So I guess let's go ahead with the current solution (which we know is at least confined to this block only). My apologies for resisting at first, I thought it'd be easier to find a solution that wouldn't involve a deprecation.

File: packages/blocks/src/api/parser/convert-legacy-block.js
Comment (ockham) [inline packages/blocks/src/api/parser/convert-legacy-block.js:64]: Let's maybe document that this alone isn't enough:

```suggestion
		name = 'core/comments';
		// Note that we also had to add a deprecation to the block in order to remove the old,
		// generated `wp-block-comments-query-loop` class name.
```

**Review by ockham — APPROVED** [review]
Let's add [this note](https://github.com/WordPress/gutenberg/pull/40506/files#r861738042), rebase, and then merge this 😄

**Review by ockham — CHANGES_REQUESTED** [review]
Spoke too soon 🤦 

While this changes the block name, it doesn't seem to affect the block ID anymore 😕 

```
<!-- wp:comments-query-loop -->
```

(Try inserting a Comments Query Loop block on `trunk`, then switching to this branch, reload the Site Editor, and switch to the Code Editor)

**Comment by luisherranz** [comment]
That was working for me when Michal solved the validation problem, but I'll check again.

What is not working is the deprecation when using a custom class name. I'll push a commit for that.

**Comment by luisherranz** [comment]
> While this changes the block name, it doesn't seem to affect the block ID anymore
> Try inserting a Comments Query Loop block on trunk, then switching to this branch, reload the Site Editor, and switch to the Code Editor

@ockham this is still working fine here. Could you please check it again?

https://user-images.githubusercontent.com/3305402/165947535-2dac18e5-481c-458d-b5c9-c9c293dfbd3e.mov

**Comment by ockham** [comment]
I'm still seeing the same issue 😕 

FWIW, I'm using the Site Editor (Single Post template). Can you try that one?

I'll restart my `npm run dev`, for good measure...

**Comment by luisherranz** [comment]
It's true that in the Site Editor I need to save the template to see the change. Not sure if that is a bug or not, and I wonder why the Site Editor and Post Editor update the Code Editor differently.

**Comment by michalczaplinski** [comment]
Hi @youknowriad Would you know why in the **site** editor, one has to save the template in order for the renamed block to appear in the editor? 

In the post editor, the updated state is reflected immediately.

https://user-images.githubusercontent.com/5417266/166084040-e401e485-d435-4a95-952a-a2df378ab054.mp4

**Comment by ockham** [comment]
Thanks all. In order to de-risk the rename a bit, I've taken the commits that change the block name (but not the ID) and moved them to a separate PR (#40763) which I hope will be included in the RC1 tomorrow 🙂

**Comment by ockham** [comment]
Rebased to reflect that https://github.com/WordPress/gutenberg/pull/40763 has been merged.

**Comment by youknowriad** [comment]
> Would you know why in the site editor, one has to save the template in order for the renamed block to appear in the editor?

AFAIK, in both cases (side editor and post editor), the deprecations run on load in the editor but in both cases if you want the updated markup in the frontend, you need to save again.

**Comment by ockham** [comment]
> > Would you know why in the site editor, one has to save the template in order for the renamed block to appear in the editor?
> 
> AFAIK, in both cases (side editor and post editor), the deprecations run on load in the editor but in both cases if you want the updated markup in the frontend, you need to save again.

I just tried again with a Comments Query Loop block in both the Post and the Site Editor. In the Post Editor, it really gets converted to a Comments block right upon loading (verified in the Code Editor); the same isn't true for the Site Editor.

Maybe this difference is circumstantial (and unwanted) -- it could be a result of how block validation and/or deprecations are run in each editor... 🤔

**Comment by ockham** [comment]
As mentioned in https://github.com/WordPress/gutenberg/pull/41807#issuecomment-1160449503, let's revisit this PR and get it in before #41807, as it should help break down complexity quite a bit.

File: packages/blocks/src/api/serializer.js
Comment (cbravobernal) [inline packages/blocks/src/api/serializer.js]: Added this line in case anyone wants to debug and discover why is it updating the code with the rename on the post editor but not on the site editor. As is mentioned [here] (https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1122391822)

I will keep taking a look at it tomorrow ( 21st June, for the async 😄  )

Comment (ockham) [reply packages/blocks/src/api/serializer.js]: Thanks Carlos! FWIW, I think it'd be fine to consider this PR for merging even if the Site Editor isn't updating the block IDs. @DAreRodz and I were running into the same issue today while working on https://github.com/WordPress/gutenberg/pull/41807, so I think it's pretty much pre-existing and unrelated to what we're doing here. We might consider fixing it, but maybe in a separate PR so we don't block this one 🙂

Comment (cbravobernal) [reply packages/blocks/src/api/serializer.js]: Comment reverted!

File: test/integration/fixtures/blocks/core__comments.serialized.html
Comment (cbravobernal) [inline test/integration/fixtures/blocks/core__comments.serialized.html]: Will this change affect anyone that relied on the class for giving styles? What is our approach in similar issues?

**Comment by cbravobernal** [comment]
I guess then the PR is ready to merge 😄

File: test/integration/fixtures/blocks/core__comments.serialized.html
Comment (ockham) [reply test/integration/fixtures/blocks/core__comments.serialized.html]: I think we need to keep the old class name around, unfortunately 😬

File: packages/block-library/src/comments/save.js
Comment (cbravobernal) [inline packages/block-library/src/comments/save.js]: Is this the best way to add a previous className for backward compatibility?

**Comment by c4rl0sbr4v0** [comment]
Like [it was previously done with the Query Loop block](https://github.com/WordPress/gutenberg/pull/32514/files#diff-4a9213eb14b1addcc5dfd9ed42344752830e90c845917fde7170b1429c7db4bcR1-R6), I'm reverting the last changes in order not to add the previous class.

**Comment by ockham** [comment]
I just tried setting some properties for the Comments block in the post editor on `trunk` (text color, resulting in `has-cyan-bluish-gray-color has-text-color`). After switching to this PR and reloading the post in the editor, the block conversion (and deprecation) runs; but the class names are duplicated on the `<div />`:

image

The resulting block in the post content:

```
<!-- wp:comments {"textColor":"cyan-bluish-gray","className":"has-cyan-bluish-gray-color has-text-color"} -->
<div class="wp-block-comments wp-block-comments-query-loop has-cyan-bluish-gray-color has-text-color has-cyan-bluish-gray-color has-text-color"><!-- wp:comments-title /-->
```

(etc)

**Comment by ockham** [comment]
Unfortunately, [this behavior](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1164632563) -- in combination with the Site Editor not updating the post content right away -- can produce a conflict:

- On `trunk`, I edited the "Single" template in the Site Editor to add the "Comments Query Loop" block, and changed its text color to "bluish gray".
- I then switched to this branch, and reloaded the template in the Site Editor. ([As we know](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1122391822), this doesn't immediately update the post content, as is visible in the Code Editor.)
- I then changed the text color to the "foreground color" (the black that's first in the "Default" palette), and saved the template.
- In the Code Editor, I noticed that the block's `<div />` now has conflicting classes 😕 

```
<!-- wp:comments {"textColor":"foreground","className":"has-cyan-bluish-gray-color has-text-color"} -->
<div class="wp-block-comments wp-block-comments-query-loop has-cyan-bluish-gray-color has-text-color has-foreground-color has-text-color"><!-- wp:comments-title /-->
```

**Comment by ockham** [comment]
Since it now seems unlikely that this PR will make it into WP 6.0.1 (see e.g. https://github.com/WordPress/gutenberg/pull/40763#issuecomment-1161743270 or https://github.com/WordPress/gutenberg/pull/41807#issuecomment-1164560396), it probably makes sense to investigate and fix the [issue](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1122391822) with the Site Editor not applying block conversions and/or deprecations to the template content before we merge this PR 😅

**Comment by ockham** [comment]
Rebased.

File: packages/block-library/src/comments/save.js
Comment (ockham) [reply packages/block-library/src/comments/save.js]: This will add it to _all_ Comments blocks, including newly inserted ones.

Overall, I think I'd prefer if we only added the old class name to blocks that are converted from the old ID -- i.e. somewhere in the deprecation (or migration).

However, I've spent half of the afternoon experimenting with that approach and still haven't found a solution that doesn't cause a block validation error 😅

**Comment by ockham** [comment]
BTW I now believe that [this problem](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1164655191) (colliding class names) is unrelated to the bug I'm fixing in #42081. I now think that it's due to our block deprecation, which accidentally duplicates class names that are added by the block-supports mechanism, and thus end up in a separate `className` block attribute (when they shouldn't).

I've debugged this a bit and will continue to update this PR with new findings (or a fix).

File: packages/block-library/src/comments/save.js
Comment (ockham) [reply packages/block-library/src/comments/save.js]: This seems to do the trick: f937d9f

File: test/integration/fixtures/blocks/core__comments__deprecated-1.json
Comment (ockham) [inline test/integration/fixtures/blocks/core__comments__deprecated-1.json]: This is to illustrate [this problem](https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1174074568).

The block-supports generated classes shouldn't be here, i.e. we'd like the following instead:

```suggestion
			"className": "comments-post-extra",
```

We'll want to change the deprecation accordingly.

File: test/integration/fixtures/blocks/core__comments__deprecated-1.serialized.html
Comment (ockham) [inline test/integration/fixtures/blocks/core__comments__deprecated-1.serialized.html]: Compare this line to the [original input](https://github.com/WordPress/gutenberg/pull/40506/files#diff-556db1e883beda2196365535b3d8872222aa469a91a3b67bca6e7a030c3fdd3eR1).

The block-supports generated classes shouldn't be here, i.e. we'd like the following instead:

```suggestion
<!-- wp:comments {"textColor":"luminous-vivid-orange","className":"comments-post-extra"} -->
```

We'll want to change the deprecation accordingly.

See https://github.com/WordPress/gutenberg/pull/40506#issuecomment-1174074568 for more context.

**Comment by ockham** [comment]
Rebased to include #42081.

**Comment by ockham** [comment]
This should be ready for another look. (I've updated the Testing Instructions in the PR description to cover all the weird little edge cases I ran into when testing this.)

**Review by DAreRodz — COMMENTED** [review]
I haven't followed the test instructions yet, but the code looks really neat.

Great job, @ockham (and others)! 👏👏

**Review by ockham — APPROVED** [review]

**Comment by DAreRodz** [comment]
I've tested it, and it works as expected. 👍

**Comment by DAreRodz** [comment]
As the PR is already approved, I'll merge it so I can rebase https://github.com/WordPress/gutenberg/pull/41807 on top of these changes.

**Comment by mburridge** [comment]
Added the Needs Dev Note label in case this needs a dev note (either individual or as part of a "misc" dev note) for WP 6.1 release.

**Pull request merged**
